File: blinky-things/generators/Water.cpp
#include "Water.h"
#include "../physics/PhysicsContext.h"
#include "../physics/MatrixBackground.h"
#include "../physics/LinearBackground.h"
#include "../physics/EdgeSpawnRegion.h"
#include "../physics/RandomSpawnRegion.h"
#include "../physics/KillBoundary.h"
#include "../physics/WrapBoundary.h"
#include "../physics/MatrixForceAdapter.h"
#include "../physics/LinearForceAdapter.h"
#include <Arduino.h>

Water::Water() : params_(), noiseTime_(0.0f), background_(nullptr) {}

bool Water::begin(const DeviceConfig& config) {
    if (!ParticleGenerator::begin(config)) return false;

    noiseTime_ = 0.0f;

    return true;
}

void Water::initPhysicsContext() {
    // Set physics parameters from WaterParams (dimension-scaled)
    gravity_ = params_.gravity * traversalDim_;
    drag_ = params_.drag;

    // Create layout-appropriate physics components
    bool wrap = PhysicsContext::shouldWrapByDefault(layout_);

    // Spawn region: top edge for matrix, random for linear
    spawnRegion_ = PhysicsContext::createSpawnRegion(
        layout_, GeneratorType::WATER, width_, height_, spawnBuffer_);

    // Boundary: kill for matrix (splash handled separately), wrap for linear
    boundary_ = PhysicsContext::createBoundary(
        layout_, GeneratorType::WATER, wrap, boundaryBuffer_);

    // Force adapter: 2D for matrix, 1D for linear
    forceAdapter_ = PhysicsContext::createForceAdapter(layout_, forceBuffer_);
    if (forceAdapter_) {
        forceAdapter_->setWind(params_.windBase, scaledWindVar());
    }

    // Background model: water surface with height variation for matrix, uniform for linear
    background_ = PhysicsContext::createBackground(
        layout_, BackgroundStyle::WATER, backgroundBuffer_);
}

void Water::generate(PixelMatrix& matrix, const AudioControl& audio) {
    // Advance noise animation time (blend between organic and music-driven)
    float organicSpeed = 0.012f + 0.008f * audio.energy;
    float musicSpeed = 0.03f + 0.02f * audio.energy;
    float timeSpeed = organicSpeed * (1.0f - audio.rhythmStrength) + musicSpeed * audio.rhythmStrength;

    // Phase-driven wave speed breathing: water flows faster on-beat
    float phaseBreathing = 0.5f + 0.5f * audio.phaseToPulse();
    float blendedBreathing = 1.0f * (1.0f - audio.rhythmStrength) + phaseBreathing * audio.rhythmStrength;
    noiseTime_ += timeSpeed * blendedBreathing;

    // Render noise background first (tropical sea underlayer)
    if (background_) {
        background_->setIntensity(params_.backgroundIntensity);
        background_->render(matrix, width_, height_, noiseTime_, audio);
    }

    // --- Depth gradient (matrix only): darken lower rows for depth perception ---
    if (PhysicsContext::isPrimaryAxisVertical(layout_) && height_ > 2) {
        for (int y = 0; y < height_; y++) {
            // Surface (y=0) is brightest, bottom is darkest
            float depthFade = 1.0f - 0.6f * ((float)y / (float)(height_ - 1));
            for (int x = 0; x < width_; x++) {
                RGB px = matrix.getPixel(x, y);
                matrix.setPixel(x, y,
                    (uint8_t)(px.r * depthFade),
                    (uint8_t)(px.g * depthFade),
                    (uint8_t)(px.b * depthFade));
            }
        }
    }

    // --- Ripple simulation (2-buffer demoscene algorithm) ---
    int rippleSize = min((int)numLeds_, (int)MAX_RIPPLE_LEDS);
    if (rippleSize > 2) {
        int16_t* curr = rippleFlip_ ? rippleA_ : rippleB_;
        int16_t* prev = rippleFlip_ ? rippleB_ : rippleA_;

        // Propagate ripples: new[i] = avg(neighbors) - current[i], with damping
        for (int i = 1; i < rippleSize - 1; i++) {
            int16_t avg = (prev[i - 1] + prev[i + 1]) >> 1;
            curr[i] = avg - curr[i];
            curr[i] -= curr[i] >> 5;  // ~3% damping per frame
        }
        rippleFlip_ = !rippleFlip_;

        // Inject ripple impulse on pulse (kick/snare → splash creates ripple)
        if (audio.pulse > 0.3f) {
            int pos = random(1, rippleSize - 1);
            curr[pos] = (int16_t)(audio.pulse * 500);
        }
        // Stronger impulse on beat
        if (beatHappened() && audio.rhythmStrength > 0.3f) {
            // Edge sweep: inject along one edge
            for (int i = 0; i < min(3, rippleSize); i++) {
                curr[i] = (int16_t)(audio.energy * 800);
            }
        }

        // Composite ripples as additive blue-cyan brightness modulation
        for (int i = 0; i < rippleSize; i++) {
            int16_t r = curr[i];
            if (r > 0) {
                uint8_t brightness = (uint8_t)min(80, (int)(r >> 2));
                int x = i % width_;
                int y = i / width_;
                if (x < width_ && y < height_) {
                    RGB px = matrix.getPixel(x, y);
                    matrix.setPixel(x, y,
                        px.r,
                        min(255, px.g + brightness),
                        min(255, px.b + (brightness >> 1)));
                }
            }
        }
    }

    // --- Bioluminescence: blue-green glow on pulse events ---
    {
        int glowSize = min((int)numLeds_, (int)MAX_RIPPLE_LEDS);

        // Decay existing glow (~5% per frame)
        for (int i = 0; i < glowSize; i++) {
            if (glowBuffer_[i] > 0) {
                glowBuffer_[i] = (uint8_t)(glowBuffer_[i] * 0.95f);
                if (glowBuffer_[i] < 2) glowBuffer_[i] = 0;
            }
        }

        // Spawn glow on pulse
        if (audio.pulse > 0.4f) {
            int pos = random(glowSize);
            uint8_t intensity = (uint8_t)(audio.pulse * audio.energy * 200);
            glowBuffer_[pos] = max(glowBuffer_[pos], intensity);
            // Spread to neighbors
            if (pos > 0) glowBuffer_[pos - 1] = max(glowBuffer_[pos - 1], (uint8_t)(intensity >> 1));
            if (pos < glowSize - 1) glowBuffer_[pos + 1] = max(glowBuffer_[pos + 1], (uint8_t)(intensity >> 1));
        }

        // Render glow as additive blue-green
        for (int i = 0; i < glowSize; i++) {
            if (glowBuffer_[i] > 5) {
                int x = i % width_;
                int y = i / width_;
                if (x < width_ && y < height_) {
                    uint8_t g = glowBuffer_[i];
                    uint8_t b = (uint8_t)(g * 0.6f);
                    RGB px = matrix.getPixel(x, y);
                    matrix.setPixel(x, y,
                        px.r,
                        min(255, px.g + g),
                        min(255, px.b + b));
                }
            }
        }
    }

    // Run particle system (spawns, updates, renders particles)
    ParticleGenerator::generate(matrix, audio);
}

void Water::reset() {
    ParticleGenerator::reset();
    noiseTime_ = 0.0f;
    memset(rippleA_, 0, sizeof(rippleA_));
    memset(rippleB_, 0, sizeof(rippleB_));
    memset(glowBuffer_, 0, sizeof(glowBuffer_));
    rippleFlip_ = false;
}

void Water::spawnParticles(float dt) {
    float spawnProb;
    uint8_t dropCount = 0;

    // MUSIC-DRIVEN behavior (rhythmStrength weighted)
    float phasePulse = audio_.phaseToPulse();
    float phaseWave = 0.4f + 0.6f * phasePulse;
    float musicSpawnProb = params_.baseSpawnChance * phaseWave + params_.audioSpawnBoost * audio_.pulse * phasePulse;

    // Wave burst on beat (scales with rhythmStrength, device width, and onset density)
    // Onset density scales burst: dance (4-6/s) = full wave, ambient (0-1/s) = gentle drip
    if (beatHappened() && audio_.rhythmStrength > 0.3f) {
        float densityScale = 0.5f + 0.5f * min(audio_.onsetDensity / 6.0f, 1.0f);
        uint8_t waveDrops = (uint8_t)min(8.0f, max(2.0f, 0.5f * crossDim_ * (1.0f + audio_.rhythmStrength)));
        dropCount += (uint8_t)(waveDrops * (0.5f + 0.5f * audio_.energy) * audio_.rhythmStrength * densityScale);
    }

    // ORGANIC-DRIVEN behavior (inverse rhythmStrength weighted)
    float smoothEnergy = 0.4f + 0.3f * audio_.energy;
    float organicSpawnProb = params_.baseSpawnChance * smoothEnergy;

    // Gentle transient response (only in organic mode to avoid double-triggering with beats)
    if (audio_.pulse > params_.organicTransientMin && audio_.rhythmStrength < 0.5f) {
        float transientStrength = (audio_.pulse - params_.organicTransientMin) /
                                 (1.0f - params_.organicTransientMin);
        dropCount += (uint8_t)(2 * transientStrength);  // 0-2 drops based on transient strength
    }

    // BLEND spawn probability between modes
    spawnProb = organicSpawnProb * (1.0f - audio_.rhythmStrength) +
                musicSpawnProb * audio_.rhythmStrength;

    // Random baseline spawning
    if (random(1000) < spawnProb * 1000) {
        dropCount++;
    }

    // Spawn drops using layout-aware spawn region
    uint16_t maxParts = scaledMaxParticles();
    for (uint8_t i = 0; i < dropCount && pool_.getActiveCount() < maxParts; i++) {
        float x, y;
        getSpawnPosition(x, y);

        // Get initial velocity from spawn region (dimension-scaled)
        float speed = scaledDropVelMin() +
                     random(100) * (scaledDropVelMax() - scaledDropVelMin()) / 100.0f;

        float vx, vy;
        getInitialVelocity(speed, vx, vy);

        // Add spread perpendicular to main direction (dimension-scaled)
        float spreadAmount = (random(200) - 100) * scaledDropSpread() / 100.0f;
        if (PhysicsContext::isPrimaryAxisVertical(layout_)) {
            vx += spreadAmount;
        } else {
            vy += spreadAmount * 0.3f;
        }

        // Blend velocity multiplier between organic (0.7x) and music (1.0-1.2x)
        float organicVelMult = 0.7f;
        float musicVelMult = 1.0f + 0.2f * audio_.pulse;
        float velocityMult = organicVelMult * (1.0f - audio_.rhythmStrength) +
                            musicVelMult * audio_.rhythmStrength;
        vx *= velocityMult;
        vy *= velocityMult;

        uint8_t intensity = random(params_.intensityMin, params_.intensityMax + 1);

        pool_.spawn(x, y, vx, vy, intensity, params_.defaultLifespan, 1.0f,
                   ParticleFlags::GRAVITY | ParticleFlags::WIND |
                   ParticleFlags::FADE | ParticleFlags::SPLASH);
    }
}

void Water::updateParticle(Particle* p, float dt) {
    // Check for splash on bottom collision (matrix) or wrap point (linear)
    if (p->hasFlag(ParticleFlags::SPLASH)) {
        bool shouldSplash = false;

        if (PhysicsContext::isPrimaryAxisVertical(layout_)) {
            // Matrix: splash at bottom
            shouldSplash = p->y >= height_ - 1;
        } else {
            // Linear: splash at edges (before wrap)
            shouldSplash = p->x < 0.5f || p->x >= width_ - 0.5f;
        }

        if (shouldSplash) {
            spawnSplash(p->x, p->y, p->intensity);
            pool_.kill(p);
        }
    }
}

void Water::renderParticle(const Particle* p, PixelMatrix& matrix) {
    int x = (int)p->x;
    int y = (int)p->y;

    if (x >= 0 && x < width_ && y >= 0 && y < height_) {
        uint32_t color = particleColor(p->intensity);
        uint8_t r = (color >> 16) & 0xFF;
        uint8_t g = (color >> 8) & 0xFF;
        uint8_t b = color & 0xFF;

        // MAX BLENDING: Drops take brightest value to pop against dark background
        RGB existing = matrix.getPixel(x, y);
        matrix.setPixel(x, y,
                       max(existing.r, r),
                       max(existing.g, g),
                       max(existing.b, b));
    }
}

uint32_t Water::particleColor(uint8_t intensity) const {
    // White/light cyan drops for visibility against blue background
    uint8_t white = intensity;
    uint8_t blue = (uint8_t)min(255, (int)intensity + 40);
    uint8_t green = (uint8_t)((uint16_t)intensity * 3 / 4);
    return ((uint32_t)white << 16) | ((uint32_t)green << 8) | blue;
}

void Water::spawnSplash(float x, float y, uint8_t parentIntensity) {
    // Calculate available slots (dimension-scaled)
    uint16_t maxParts = scaledMaxParticles();
    uint8_t available = maxParts > pool_.getActiveCount()
                        ? maxParts - pool_.getActiveCount()
                        : 0;
    uint8_t splashCount = min(scaledSplashParticles(), available);

    // Guard against division by zero in angle calculation
    if (splashCount == 0) return;

    for (uint8_t i = 0; i < splashCount; i++) {
        // Radial splash pattern
        float angle = (i * TWO_PI / splashCount) + random(100) * 0.01f;
        float speed = scaledSplashVelMin() +
                     random(100) * (scaledSplashVelMax() - scaledSplashVelMin()) / 100.0f;

        float vx = cos(angle) * speed;
        float vy = sin(angle) * speed - 1.0f;  // Slight upward component

        uint8_t intensity = (uint8_t)min((uint32_t)255, (uint32_t)parentIntensity * params_.splashIntensity / 255);

        pool_.spawn(x, y, vx, vy, intensity, 30, 0.5f,
                   ParticleFlags::GRAVITY | ParticleFlags::FADE);
    }
}

File: blinky-simulator/ts-src/generators/Water.ts
/**
 * Water Generator - Port of blinky-things/generators/Water.cpp
 *
 * Simulates water drops falling and creating splashes.
 * Simplified version that captures the visual essence without full particle system.
 */

import { Generator, AudioControl, PixelMatrix, RGB, createPixelMatrix, AudioControlHelpers } from '../types';

/**
 * Water color palette (matches firmware Palette::WATER)
 * Uses three-segment interpolation: black -> medium blue -> cyan -> light blue
 */
function intensityToColor(intensity: number): RGB {
  // Clamp to 0-255
  intensity = Math.max(0, Math.min(255, Math.floor(intensity * 255)));

  if (intensity === 0) {
    return { r: 0, g: 0, b: 0 };
  } else if (intensity < 85) {
    // Black to medium blue
    const t = intensity / 84;
    return {
      r: 0,
      g: 0,
      b: Math.round(150 * t)
    };
  } else if (intensity < 170) {
    // Medium blue to cyan
    const t = (intensity - 85) / 84;
    return {
      r: 0,
      g: Math.round(120 * t),
      b: Math.round(150 + (255 - 150) * t)
    };
  } else {
    // Cyan to light blue
    const t = (intensity - 170) / 85;
    return {
      r: Math.round(80 * t),
      g: Math.round(120 + (200 - 120) * t),
      b: 255
    };
  }
}

interface Drop {
  x: number;
  y: number;
  vx: number;
  vy: number;
  intensity: number;
  active: boolean;
}

interface Splash {
  x: number;
  y: number;
  vx: number;
  vy: number;
  intensity: number;
  age: number;
  active: boolean;
}

export interface WaterParams {
  baseSpawnChance: number;   // Drop spawn probability (default: 0.05)
  audioSpawnBoost: number;   // Audio boost for spawning (default: 0.4)
  dropVelocityMin: number;   // Minimum drop fall speed (default: 0.5)
  dropVelocityMax: number;   // Maximum drop fall speed (default: 1.5)
  dropSpread: number;        // Horizontal velocity variation (default: 0.3)
  gravity: number;           // Gravity strength (default: 0.3)
  splashParticles: number;   // Particles per splash (default: 4)
  maxParticles: number;      // Maximum concurrent particles (default: 24)
}

export class WaterGenerator implements Generator {
  name = 'Water';

  private width = 0;
  private height = 0;
  private matrix: PixelMatrix | null = null;
  private lastTimeMs = -1;
  private lastPulse = 0;

  private drops: Drop[] = [];
  private splashes: Splash[] = [];

  private params: WaterParams = {
    baseSpawnChance: 0.05,
    audioSpawnBoost: 0.4,
    dropVelocityMin: 0.5,
    dropVelocityMax: 1.5,
    dropSpread: 0.3,
    gravity: 0.3,
    splashParticles: 4,
    maxParticles: 24
  };

  begin(width: number, height: number): void {
    this.width = width;
    this.height = height;
    this.matrix = createPixelMatrix(width, height);
    this.drops = [];
    this.splashes = [];
    this.lastTimeMs = -1;
    this.lastPulse = 0;
  }

  setParams(params: Partial<WaterParams>): void {
    this.params = { ...this.params, ...params };
  }

  private beatHappened(audio: AudioControl): boolean {
    const happened = audio.pulse > 0.5 && this.lastPulse <= 0.5;
    this.lastPulse = audio.pulse;
    return happened;
  }

  private spawnSplash(x: number, y: number, parentIntensity: number): void {
    const { splashParticles, maxParticles } = this.params;
    const totalParticles = this.drops.filter(d => d.active).length + this.splashes.filter(s => s.active).length;

    const available = Math.max(0, maxParticles - totalParticles);
    const count = Math.min(splashParticles, available);

    for (let i = 0; i < count; i++) {
      const angle = (i * Math.PI * 2 / count) + Math.random() * 0.1;
      const speed = 0.3 + Math.random() * 0.3;

      this.splashes.push({
        x,
        y,
        vx: Math.cos(angle) * speed,
        vy: Math.sin(angle) * speed - 0.5,  // Slight upward component
        intensity: parentIntensity * 0.7,
        age: 0,
        active: true
      });
    }
  }

  update(audio: AudioControl, timeMs: number): void {
    if (!this.matrix) return;

    // Handle first frame
    if (this.lastTimeMs < 0) {
      this.lastTimeMs = timeMs;
    }

    const deltaMs = Math.min(timeMs - this.lastTimeMs, 100);
    this.lastTimeMs = timeMs;

    if (deltaMs < 16) return;

    const dt = deltaMs / 33;  // Normalize to ~30fps
    const { baseSpawnChance, audioSpawnBoost, dropVelocityMin, dropVelocityMax, dropSpread, gravity, maxParticles } = this.params;
    const hasRhythm = AudioControlHelpers.hasRhythm(audio);

    // Clear matrix
    this.matrix.clear();

    // 1. Spawn new drops
    let dropCount = 0;
    let spawnProb = baseSpawnChance;

    if (hasRhythm) {
      // Music mode: beat-synced spawning
      if (audio.pulse > 0.3) {
        const beatBoost = AudioControlHelpers.phaseToPulse(audio);
        spawnProb += audioSpawnBoost * audio.pulse * beatBoost;
      }

      if (this.beatHappened(audio)) {
        dropCount = 4;  // Wave on beat
      }
    } else {
      // Organic mode: transient-reactive
      if (audio.pulse > 0.5) {
        spawnProb += audioSpawnBoost * audio.pulse;
        dropCount = 2;
      }
    }

    if (Math.random() < spawnProb) {
      dropCount++;
    }

    // Spawn drops from top
    const totalParticles = this.drops.filter(d => d.active).length + this.splashes.filter(s => s.active).length;
    for (let i = 0; i < dropCount && totalParticles + i < maxParticles; i++) {
      const vy = dropVelocityMin + Math.random() * (dropVelocityMax - dropVelocityMin);
      const vx = (Math.random() - 0.5) * 2 * dropSpread;
      const intensity = 0.6 + Math.random() * 0.4;

      this.drops.push({
        x: Math.random() * this.width,
        y: 0,
        vx,
        vy,
        intensity,
        active: true
      });
    }

    // 2. Update drops
    for (const drop of this.drops) {
      if (!drop.active) continue;

      drop.vy += gravity * dt;
      drop.x += drop.vx * dt;
      drop.y += drop.vy * dt;

      // Splash on bottom collision
      if (drop.y >= this.height - 1) {
        this.spawnSplash(drop.x, this.height - 1, drop.intensity);
        drop.active = false;
      }
    }

    // 3. Update splashes
    for (const splash of this.splashes) {
      if (!splash.active) continue;

      splash.vy += gravity * dt;
      splash.x += splash.vx * dt;
      splash.y += splash.vy * dt;
      splash.age += dt;
      splash.intensity *= 0.95;  // Fade

      // Deactivate when faded or out of bounds
      if (splash.intensity < 0.05 || splash.age > 30 ||
          splash.y >= this.height || splash.y < 0 ||
          splash.x < 0 || splash.x >= this.width) {
        splash.active = false;
      }
    }

    // 4. Render drops
    for (const drop of this.drops) {
      if (!drop.active) continue;

      const x = Math.floor(drop.x);
      const y = Math.floor(drop.y);
      if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
        const color = intensityToColor(drop.intensity);
        const existing = this.matrix.getPixel(x, y);
        // Additive blend
        this.matrix.setPixel(x, y, {
          r: Math.min(255, existing.r + color.r),
          g: Math.min(255, existing.g + color.g),
          b: Math.min(255, existing.b + color.b)
        });
      }
    }

    // 5. Render splashes
    for (const splash of this.splashes) {
      if (!splash.active) continue;

      const x = Math.floor(splash.x);
      const y = Math.floor(splash.y);
      if (x >= 0 && x < this.width && y >= 0 && y < this.height) {
        const color = intensityToColor(splash.intensity);
        const existing = this.matrix.getPixel(x, y);
        this.matrix.setPixel(x, y, {
          r: Math.min(255, existing.r + color.r),
          g: Math.min(255, existing.g + color.g),
          b: Math.min(255, existing.b + color.b)
        });
      }
    }

    // 6. Add base water ambient (subtle wave)
    for (let y = 0; y < this.height; y++) {
      for (let x = 0; x < this.width; x++) {
        const existing = this.matrix.getPixel(x, y);
        if (existing.r === 0 && existing.g === 0 && existing.b === 0) {
          const wave = Math.sin((x + y * 0.5 + timeMs / 500) * 0.8) * 0.05 + 0.08;
          const color = intensityToColor(wave * (1 + audio.energy * 0.3));
          this.matrix.setPixel(x, y, color);
        }
      }
    }

    // Clean up inactive particles
    this.drops = this.drops.filter(d => d.active);
    this.splashes = this.splashes.filter(s => s.active);
  }

  getMatrix(): PixelMatrix {
    if (!this.matrix) {
      throw new Error('Water generator not initialized');
    }
    return this.matrix;
  }

  reset(): void {
    this.drops = [];
    this.splashes = [];
    this.matrix?.clear();
    this.lastTimeMs = -1;
    this.lastPulse = 0;
  }
}
